File: decl-py/src/decl/qengine/programs.py
"""Shared executable expressions; no operation captures a record instance."""

from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING, Any

from decl.semantics import (
    ABSENT,
    ArrV,
    Closure,
    EvalErr,
    Key,
    MapV,
    NsRef,
    Pattern,
    PreArr,
    PreObj,
    PreVal,
    Quantity,
    RecInst,
    Scope,
    Segs,
    StdRef,
    is_int,
    is_str,
    mentions_referrers,
)

if TYPE_CHECKING:
    from decl.engine import Engine

Op = Callable[["Engine", Scope], Any]


class Programs:
    def __init__(self) -> None:
        self.expressions: dict[int, tuple[Any, Op]] = {}
        self.callees: dict[int, tuple[Any, Op]] = {}
        self.navigations: dict[int, tuple[Any, Op]] = {}
        self.compiled = 0
        self.compiled_schemas = 0
        self.schemas: dict[int, tuple[Any, Any, list[Any], set[str]]] = {}

    def schema(self, rt: Any) -> tuple[list[Any], set[str]]:
        hit = self.schemas.get(id(rt))
        if hit is not None and hit[1] is rt["members"] and len(hit[2]) == len(rt["members"]):
            return hit[2], hit[3]
        plans = [
            (
                m,
                m.get("conj") or [m["type"]],
                self.get(m["expr"]) if m.get("expr") else None,
                self.get(m["dflt"]) if m.get("dflt") else None,
                mentions_referrers(m.get("expr") if m["kind"] == "der" else m.get("dflt")),
            )
            for m in rt["members"]
        ]
        names = {m["name"] for m in rt["members"]}
        self.schemas[id(rt)] = (rt, rt["members"], plans, names)
        self.compiled_schemas += 1
        return plans, names

    def get(self, e: dict[str, Any]) -> Op:
        hit = self.expressions.get(id(e))
        if hit is not None:
            return hit[1]
        op = self.compile(e)
        self.expressions[id(e)] = (e, op)
        self.compiled += 1
        return op

    def callee(self, e: dict[str, Any]) -> Op:
        hit = self.callees.get(id(e))
        if hit is not None:
            return hit[1]
        op: Op
        if e["e"] == "member":
            base, name = self.callee(e["x"]), e["name"]

            def member_callee(eng: Engine, sc: Scope) -> Any:
                x = base(eng, sc)
                if isinstance(x, StdRef):
                    return StdRef([*x.path, name])
                if isinstance(x, NsRef):
                    return eng.ns_value(x, name, sc)
                return eng.access(eng.deref(x), name)

            op = member_callee
        else:
            op = self.get(e)
        self.callees[id(e)] = (e, op)
        return op

    def nav(self, e: dict[str, Any]) -> Op:
        hit = self.navigations.get(id(e))
        if hit is not None:
            return hit[1]
        kind = e["e"]
        if kind == "paren":
            op = self.nav(e["x"])
        elif kind == "if":
            c, t, f = self.get(e["c"]), self.nav(e["t"]), self.nav(e["f"])

            def conditional(eng: Engine, sc: Scope) -> Any:
                return t(eng, sc) if eng.truthy(c(eng, sc)) else f(eng, sc)

            op = conditional
        elif kind == "member":
            base, name = self.nav(e["x"]), e["name"]

            def member(eng: Engine, sc: Scope) -> Any:
                raw = base(eng, sc)
                if isinstance(raw, Segs):
                    return Segs([*raw.segs, name])
                x = eng.deref(raw)
                v = eng.access(x, name)
                return Segs([*x.path, name]) if v is ABSENT and isinstance(x, RecInst) else v

            op = member
        elif kind == "index":
            base, index, ordinary = self.nav(e["x"]), self.get(e["i"]), self.get(e)

            def indexed(eng: Engine, sc: Scope) -> Any:
                raw, i = base(eng, sc), index(eng, sc)
                seg = int(i) if is_int(i) else Key(i)
                if isinstance(raw, Segs):
                    return Segs([*raw.segs, seg])
                x = eng.deref(raw)
                if isinstance(x, ArrV):
                    n = int(i)
                    return x.items[n] if 0 <= n < len(x.items) else Segs([*x.path, n])
                if isinstance(x, MapV):
                    return x.entries[i] if i in x.entries else Segs([*x.path, Key(i)])
                if isinstance(x, RecInst):
                    v = eng.access(x, i)
                    return Segs([*x.path, i]) if v is ABSENT else v
                return ordinary(eng, sc)

            op = indexed
        else:
            op = self.get(e)
        self.navigations[id(e)] = (e, op)
        return op

    def compile(self, e: dict[str, Any]) -> Op:
        k = e["e"]
        if k == "lit":
            v = e["v"]
            return lambda eng, sc: v
        if k == "pattern":
            return lambda eng, sc: Pattern(e["re"])
        if k == "unitlit":

            def unit(eng: Engine, sc: Scope) -> Any:
                try:
                    u = eng.env.unit_info(e["unit"])
                except RuntimeError as error:
                    raise EvalErr(str(error)) from error
                return Quantity(u["key"], e["num"] * u["to_base"])

            return unit
        if k == "paren":
            return self.get(e["x"])
        if k == "name":
            name = e["name"]
            return lambda eng, sc: eng.name_value(name, sc)
        if k == "ctx":
            name = e["name"]
            return lambda eng, sc: eng.context_value(name, sc)
        if k == "referrers":
            return lambda eng, sc: eng.referrers(e["type"], e["member"], sc)
        if k == "obj":
            entries = e["entries"]
            return lambda eng, sc: PreObj([(en["key"], PreVal(en["val"], sc)) for en in entries])
        if k == "arr":
            items = e["items"]
            return lambda eng, sc: PreArr([(it["spread"], PreVal(it["expr"], sc)) for it in items])
        if k == "spread":

            def spread(eng: Engine, sc: Scope) -> Any:
                raise EvalErr("spread outside an object literal")

            return spread
        if k == "template":
            parts = [p if is_str(p) else self.get(p) for p in e["parts"]]
            return lambda eng, sc: "".join(
                p if isinstance(p, str) else eng.to_str(p(eng, sc)) for p in parts
            )
        if k == "if":
            c, t, f = self.get(e["c"]), self.get(e["t"]), self.get(e["f"])
            return lambda eng, sc: t(eng, sc) if eng.truthy(c(eng, sc)) else f(eng, sc)
        if k == "un":
            x, operator = self.get(e["x"]), e["op"]

            def unary(eng: Engine, sc: Scope) -> Any:
                v = x(eng, sc)
                if operator == "!":
                    return not eng.truthy(v)
                if operator == "-":
                    if v is ABSENT:
                        raise EvalErr("absent consumed")
                    return Quantity(v.dim, -v.value) if isinstance(v, Quantity) else -v
                if operator == "~":
                    return ~v
                raise EvalErr("un")

            return unary
        if k == "bin":
            operator = e["op"]
            if operator == "|>":
                r = e["r"]
                return self.get(
                    {
                        "e": "call",
                        "fn": r["fn"] if r["e"] == "call" else r,
                        "args": [e["l"], *(r["args"] if r["e"] == "call" else [])],
                    }
                )
            l, r = self.get(e["l"]), self.get(e["r"])
            if operator == "&&":
                return lambda eng, sc: eng.truthy(r(eng, sc)) if eng.truthy(l(eng, sc)) else False
            if operator == "||":
                return lambda eng, sc: True if eng.truthy(l(eng, sc)) else eng.truthy(r(eng, sc))
            if operator == "??":

                def coalesce(eng: Engine, sc: Scope) -> Any:
                    v = l(eng, sc)
                    return r(eng, sc) if v is ABSENT or v is None else v

                return coalesce
            return lambda eng, sc: eng.apply_bin(operator, l(eng, sc), r(eng, sc))
        if k == "member":
            base, name, safe = self.get(e["x"]), e["name"], e.get("safe", False)

            def access(eng: Engine, sc: Scope) -> Any:
                v = base(eng, sc)
                if isinstance(v, NsRef):
                    return eng.ns_value(v, name, sc)
                if safe and (v is None or v is ABSENT):
                    return ABSENT
                return eng.access(eng.deref(v), name)

            return access
        if k == "index":
            base, index = self.get(e["x"]), self.get(e["i"])

            def at(eng: Engine, sc: Scope) -> Any:
                v, i = eng.mat_val(base(eng, sc)), index(eng, sc)
                if isinstance(v, ArrV):
                    n = int(i)
                    if n < 0 or n >= len(v.items):
                        raise EvalErr(f"index {n} out of bounds", "E5005")
                    return v.items[n]
                if isinstance(v, MapV):
                    return v.entries.get(i, ABSENT)
                if isinstance(v, RecInst):
                    return eng.access(v, i)
                raise EvalErr("index on non-collection")

            return at
        if k == "call":
            args, fn = [self.get(a) for a in e["args"]], self.callee(e["fn"])

            def call(eng: Engine, sc: Scope) -> Any:
                values = [a(eng, sc) for a in args]
                return eng.call(fn(eng, sc), values, sc)

            return call
        if k == "lambda":
            self.get(e["body"])
            return lambda eng, sc: Closure(e["params"], e["body"], sc)
        if k == "with":
            base, patch = self.get(e["base"]), self.get(e["patch"])

            def patched(eng: Engine, sc: Scope) -> Any:
                v = eng.deref(base(eng, sc))
                if not isinstance(v, (PreObj, MapV, RecInst)):
                    raise EvalErr("with on non-record")
                return eng.with_value(v, patch(eng, sc))

            return patched
        if k == "match":
            subject = self.get(e["subject"])
            arms = [{**a, "op": self.get(a["body"])} for a in e["arms"]]

            def match(eng: Engine, sc: Scope) -> Any:
                v = eng.deref(subject(eng, sc))

                def run(arm: Any) -> Any:
                    locals_ = dict(sc.locals)
                    locals_[arm["v"]] = v
                    return arm["op"](eng, sc.with_locals(locals_))

                fallback = None
                for arm in arms:
                    if arm.get("type") is None:
                        fallback = arm
                    elif eng.member_of(v, (sc.menv or eng.env).resolve(arm["type"]), sc):
                        return run(arm)
                if fallback is not None:
                    return run(fallback)
                raise EvalErr("match: no arm matched")

            return match
        if k in ("comp", "mapcomp"):
            clauses = [
                (c["v"], self.get(c["iter"]), [self.get(f) for f in c["filters"]])
                for c in e["clauses"]
            ]
            if k == "comp":
                self.get(e["head"])
            key = self.get(e["key"]) if k == "mapcomp" else None
            val = self.get(e["val"]) if k == "mapcomp" else None

            def comprehension(eng: Engine, sc: Scope) -> Any:
                entries: list[Any] = []
                items: list[Any] = []

                def visit(ci: int, scope: Scope) -> None:
                    if ci == len(clauses):
                        if k == "comp":
                            items.append((False, PreVal(e["head"], scope)))
                        else:
                            assert key is not None and val is not None
                            n = key(eng, scope)
                            if not is_str(n):
                                raise EvalErr("map key must be string")
                            if any(kk == n for kk, _ in entries):
                                raise EvalErr(f"duplicate key {n}", "E5004")
                            entries.append((n, val(eng, scope)))
                        return
                    name, iterator, filters = clauses[ci]
                    for item in eng.iterate(iterator(eng, scope)):
                        locals_ = dict(scope.locals)
                        locals_[name] = item
                        nxt = scope.with_locals(locals_)
                        if all(eng.truthy(f(eng, nxt)) for f in filters):
                            visit(ci + 1, nxt)

                visit(0, sc)
                return PreArr(items) if k == "comp" else PreObj(entries)

            return comprehension
        raise EvalErr(f"ev: unhandled {k}")

File: decl-ts/src/qengine/programs.ts
// Executable expressions are shared by syntax identity. Operations receive the
// engine and lexical frame at execution time; none capture a record instance.
import type { Expr } from '../ast.ts';
import type { Engine, Scope } from '../engine.ts';
import { ABSENT, EvalErr, isArr, isMap, isRec, mapKey } from '../semantics.ts';
import type { Value, RT } from '../semantics.ts';

export type Op = (eng: Engine, scope: Scope) => Value;

export interface MemberPlan {
  member: RT;
  types: RT[];
  expression?: Op;
  fallback?: Op;
  deferred: boolean;
}
interface Schema {
  source: RT[];
  members: MemberPlan[];
  names: Set<string>;
}

export class Programs {
  private readonly expressions = new WeakMap<Expr, Op>();
  private readonly callees = new WeakMap<Expr, Op>();
  private readonly navigations = new WeakMap<Expr, Op>();
  private readonly schemas = new WeakMap<RT, Schema>();
  compiled = 0;
  compiledSchemas = 0;

  schema(rt: RT, mentions: (e: Expr) => boolean): Schema {
    const hit = this.schemas.get(rt);
    if (hit && hit.source === rt.members && hit.members.length === rt.members.length) return hit;
    const members = rt.members.map((m: RT): MemberPlan => ({
      member: m,
      types: m.conj ?? [m.type],
      expression: m.expr ? this.get(m.expr) : undefined,
      fallback: m.dflt ? this.get(m.dflt) : undefined,
      deferred: mentions(m.kind === 'der' ? m.expr : m.dflt),
    }));
    const schema = {
      source: rt.members,
      members,
      names: new Set<string>(members.map((m: MemberPlan) => m.member.name)),
    };
    this.schemas.set(rt, schema);
    this.compiledSchemas++;
    return schema;
  }

  get(e: Expr): Op {
    let op = this.expressions.get(e);
    if (!op) {
      op = this.compile(e);
      this.expressions.set(e, op);
      this.compiled++;
    }
    return op;
  }

  callee(e: Expr): Op {
    let op = this.callees.get(e);
    if (!op) {
      if (e.e === 'member') {
        const base = this.callee(e.x),
          name = e.name;
        op = (eng, sc) => {
          const x = base(eng, sc);
          if (x?.__std) return { __std: true, path: [...x.path, name] };
          if (x?.__nsref) return eng.nsValue(x, name, sc);
          return eng.access(eng.deref(x), name);
        };
      } else op = this.get(e);
      this.callees.set(e, op);
    }
    return op;
  }

  nav(e: Expr): Op {
    let op = this.navigations.get(e);
    if (op) return op;
    if (e.e === 'paren') op = this.nav(e.x);
    else if (e.e === 'if') {
      const c = this.get(e.c),
        t = this.nav(e.t),
        f = this.nav(e.f);
      op = (eng, sc) => (eng.truthy(c(eng, sc)) ? t(eng, sc) : f(eng, sc));
    } else if (e.e === 'member') {
      const base = this.nav(e.x),
        name = e.name;
      op = (eng, sc) => {
        const raw = base(eng, sc);
        if (raw?.__segs) return { __segs: [...raw.__segs, name] };
        const x = eng.deref(raw),
          v = eng.access(x, name);
        return v === ABSENT && isRec(x) ? { __segs: [...x.path, name] } : v;
      };
    } else if (e.e === 'index') {
      const base = this.nav(e.x),
        index = this.get(e.i),
        ordinary = this.get(e);
      op = (eng, sc) => {
        const raw = base(eng, sc),
          i = index(eng, sc);
        const seg = typeof i === 'bigint' ? Number(i) : mapKey(i);
        if (raw?.__segs) return { __segs: [...raw.__segs, seg] };
        const x = eng.deref(raw);
        if (isArr(x)) return x.items[Number(i)] ?? { __segs: [...x.path, Number(i)] };
        if (isMap(x)) return x.entries.get(i) ?? { __segs: [...x.path, mapKey(i)] };
        if (isRec(x)) {
          const v = eng.access(x, i);
          return v === ABSENT ? { __segs: [...x.path, i] } : v;
        }
        return ordinary(eng, sc);
      };
    } else op = this.get(e);
    this.navigations.set(e, op);
    return op;
  }

  private compile(e: Expr): Op {
    switch (e.e) {
      case 'lit':
        return () => e.v;
      case 'pattern':
        return () => ({ __pat: true, re: e.re });
      case 'unitlit':
        return (eng) => {
          let u: { key: string; toBase: number };
          try {
            u = eng.env.unitInfo(e.unit);
          } catch (error) {
            throw new EvalErr((error as Error).message);
          }
          return { __q: true, dim: u.key, value: e.num * u.toBase };
        };
      case 'paren':
        return this.get(e.x);
      case 'name':
        return (eng, sc) => eng.nameValue(e.name, sc);
      case 'ctx':
        return (eng, sc) => eng.contextValue(e.name, sc);
      case 'referrers':
        return (eng, sc) => eng.referrers(e.type, e.member, sc);
      case 'obj': {
        const entries = e.entries;
        return (_eng, sc) => ({
          __pre: 'obj',
          entries: entries.map((en) => [en.key, { __expr: en.val, scope: sc }]),
        });
      }
      case 'arr': {
        const items = e.items;
        return (_eng, sc) => ({
          __pre: 'arr',
          items: items.map((it) => ({ spread: it.spread, v: { __expr: it.expr, scope: sc } })),
        });
      }
      case 'spread':
        return () => {
          throw new EvalErr('spread outside an object literal');
        };
      case 'template': {
        const parts = e.parts.map((p) => (typeof p === 'string' ? p : this.get(p)));
        return (eng, sc) => {
          let s = '';
          for (const p of parts) s += typeof p === 'string' ? p : eng.toStr(p(eng, sc));
          return s;
        };
      }
      case 'if': {
        const c = this.get(e.c),
          t = this.get(e.t),
          f = this.get(e.f);
        return (eng, sc) => (eng.truthy(c(eng, sc)) ? t(eng, sc) : f(eng, sc));
      }
      case 'un': {
        const x = this.get(e.x);
        if (e.op === '!') return (eng, sc) => !eng.truthy(x(eng, sc));
        if (e.op === '-') return (eng, sc) => -x(eng, sc);
        if (e.op === '~') return (eng, sc) => ~x(eng, sc);
        return () => {
          throw new EvalErr('un');
        };
      }
      case 'bin': {
        if (e.op === '|>')
          return this.get(
            e.r.e === 'call'
              ? { e: 'call', fn: e.r.fn, args: [e.l, ...e.r.args] }
              : { e: 'call', fn: e.r, args: [e.l] },
          );
        const l = this.get(e.l),
          r = this.get(e.r),
          op = e.op;
        if (op === '&&')
          return (eng, sc) => (eng.truthy(l(eng, sc)) ? eng.truthy(r(eng, sc)) : false);
        if (op === '||')
          return (eng, sc) => (eng.truthy(l(eng, sc)) ? true : eng.truthy(r(eng, sc)));
        if (op === '??')
          return (eng, sc) => {
            const v = l(eng, sc);
            return v === ABSENT || v === null ? r(eng, sc) : v;
          };
        return (eng, sc) => eng.applyBin(op, l(eng, sc), r(eng, sc));
      }
      case 'member': {
        const base = this.get(e.x),
          name = e.name,
          safe = e.safe;
        return (eng, sc) => {
          const x = base(eng, sc);
          if (x?.__nsref) return eng.nsValue(x, name, sc);
          if (safe && (x === null || x === ABSENT)) return ABSENT;
          return eng.access(eng.deref(x), name);
        };
      }
      case 'index': {
        const base = this.get(e.x),
          index = this.get(e.i);
        return (eng, sc) => {
          const x = eng.matVal(base(eng, sc)),
            i = index(eng, sc);
          if (isArr(x)) {
            const n = Number(i);
            if (n < 0 || n >= x.items.length)
              throw new EvalErr(`index ${n} out of bounds`, 'E5005');
            return x.items[n];
          }
          if (isMap(x)) return x.entries.has(i) ? x.entries.get(i) : ABSENT;
          if (isRec(x)) return eng.access(x, i);
          throw new EvalErr('index on non-collection');
        };
      }
      case 'call': {
        const args = e.args.map((a) => this.get(a)),
          fn = this.callee(e.fn);
        return (eng, sc) => {
          const vs = args.map((a) => a(eng, sc));
          return eng.call(fn(eng, sc), vs, sc);
        };
      }
      case 'lambda': {
        this.get(e.body);
        return (_eng, sc) => ({ __clo: true, params: e.params, body: e.body, scope: sc });
      }
      case 'with': {
        const base = this.get(e.base),
          patch = this.get(e.patch);
        return (eng, sc) => {
          const v = eng.deref(base(eng, sc));
          if (!(v?.__pre === 'obj' || isMap(v) || isRec(v)))
            throw new EvalErr('with on non-record');
          return eng.withValue(v, patch(eng, sc));
        };
      }
      case 'match': {
        const subject = this.get(e.subject),
          arms = e.arms.map((a) => ({ ...a, op: this.get(a.body) }));
        return (eng, sc) => {
          const v = eng.deref(subject(eng, sc));
          const run = (arm: (typeof arms)[number]) => {
            const locals = new Map(sc.locals);
            locals.set(arm.v, v);
            return arm.op(eng, { ...sc, locals });
          };
          let fallback: (typeof arms)[number] | undefined;
          for (const arm of arms) {
            if (!arm.type) fallback = arm;
            else if (eng.memberOf(v, (sc.menv ?? eng.env).resolve(arm.type), sc)) return run(arm);
          }
          if (fallback) return run(fallback);
          throw new EvalErr('match: no arm matched');
        };
      }
      case 'comp':
      case 'mapcomp': {
        const clauses = e.clauses.map((c) => ({
          name: c.v,
          iter: this.get(c.iter),
          filters: c.filters.map((f) => this.get(f)),
        }));
        if (e.e === 'comp') this.get(e.head);
        const key = e.e === 'mapcomp' ? this.get(e.key) : null,
          val = e.e === 'mapcomp' ? this.get(e.val) : null;
        return (eng, sc) => {
          const entries: [string, Value][] = [],
            items: Value[] = [];
          const visit = (ci: number, scope: Scope) => {
            if (ci === clauses.length) {
              if (e.e === 'comp') items.push({ spread: false, v: { __expr: e.head, scope } });
              else {
                const k = key!(eng, scope);
                if (typeof k !== 'string') throw new EvalErr('map key must be string');
                if (entries.some(([n]) => n === k))
                  throw new EvalErr(`duplicate key ${k}`, 'E5004');
                entries.push([k, val!(eng, scope)]);
              }
              return;
            }
            const c = clauses[ci];
            for (const item of eng.iterate(c.iter(eng, scope))) {
              const locals = new Map(scope.locals);
              locals.set(c.name, item);
              const next = { ...scope, locals };
              if (c.filters.every((f) => eng.truthy(f(eng, next)))) visit(ci + 1, next);
            }
          };
          visit(0, sc);
          return e.e === 'comp' ? { __pre: 'arr', items } : { __pre: 'obj', entries };
        };
      }
    }
  }
}
